static/src/javascripts/projects/commercial/modules/messenger.js

```diff
@@ -0,0 +1,240 @@
+// @flow
+import reportError from 'lib/report-error';
+import dfpOrigin from 'commercial/modules/messenger/dfp-origin';
+import postMessage from 'commercial/modules/messenger/post-message';
+
+const ALLOWED_HOSTS = [dfpOrigin, `${location.protocol}//${location.host}`];
+const LISTENERS = {};
+let REGISTERED_LISTENERS = 0;
+
+const error405 = {
+    code: 405,
+    message: 'Service %% not implemented',
+};
+const error500 = {
+    code: 500,
+    message: 'Internal server error\n\n%%',
+};
+
+type StandardMessage = {
+    id: string,
+    type: string,
+    iframeId: string,
+    value: {
+        height: number,
+        width: number,
+    },
+};
+
+// A legacy from programmatic ads running in friendly iframes. They can
+// on occasion be larger than the size returned by DFP. And so they
+// have been setup to send a message of the form:
+type Payload = {
+    type: string,
+    value: {
+        id: string,
+        height: number,
+        width: number,
+    },
+};
+
+const isProgrammaticMessage = (payload: Payload): boolean =>
+    payload.type === 'set-ad-height' &&
+    'id' in payload.value &&
+    'height' in payload.value;
+
+const toStandardMessage = (payload: Payload): StandardMessage => ({
+    id: 'aaaa0000-bb11-cc22-dd33-eeeeee444444',
+    type: 'resize',
+    iframeId: payload.value.id,
+    value: {
+        height: +payload.value.height,
+        width: +payload.value.width,
+    },
+});
+
+// Incoming messages contain the ID of the iframe into which the
+// source window is embedded.
+const getIframe = (data: Object): HTMLElement | null =>
+    document.getElementById(data.iframeId);
+
+// Until DFP provides a way for us to identify with 100% certainty our
+// in-house creatives, we are left with doing some basic tests
+// such as validating the anatomy of the payload and whitelisting
+// event type
+const isValidPayload = (payload: Object): boolean =>
+    'type' in payload &&
+    'value' in payload &&
+    'id' in payload &&
+    payload.type in LISTENERS &&
+    /^[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/.test(payload.id);
+
+// Cheap string formatting function. It accepts as its first argument
+// an object `{ code, message }`. `message` is a string where successive
+// occurences of %% will be replaced by the following arguments. e.g.
+//
+// formatError({ message: "%%, you are so %%" }, "Regis", "lovely")
+//
+// returns `{ message: "Regis, you are so lovely" }`. Oh, thank you!
+const formatError = (...args) => {
+    if (args.length < 2) {
+        return args[0] || '';
+    }
+
+    const error = args[0];
+    Array.prototype.slice.call(args, 1).forEach(arg => {
+        // Keep in mind that when the first argument is a string,
+        // String.replace only replaces the first occurence
+        error.message = error.message.replace('%%', arg);
+    });
+
+    return error;
+};
+
+const onMessage = (event: Object) => {
+    let data;
```

Comment:
Yes it does, although in a separate project I forked `messenger` and moved the `try-catch` block into its own function. That is because many browser engines de-optimise them, so it's better if they site alone in a small function. So:
```
const data = decodeData(event.data)
...
function decodeData(data) {
  try {
    return JSON.parse(data);
  } catch(_) 
    return {};
}
```